types.go

```diff
@@ -34,101 +38,51 @@ func PHPString(s string, persistent bool) unsafe.Pointer {
-	Type PHPKeyType
-	Str  string
-	Int  int64
-}
-
-// Array represents a PHP array with ordered key-value pairs
-type Array struct {
-	keys   []PHPKey
-	values []interface{}
+// AssociativeArray represents a PHP array with ordered key-value pairs
+type AssociativeArray struct {
+	Map   map[string]interface{}
+	Order []string
 }
 
-// SetInt sets a value with an integer key
-func (arr *Array) SetInt(key int64, value interface{}) {
-	arr.keys = append(arr.keys, PHPKey{Type: PHPIntKey, Int: key})
-	arr.values = append(arr.values, value)
-}
-
-// SetString sets a value with a string key
-func (arr *Array) SetString(key string, value interface{}) {
-	arr.keys = append(arr.keys, PHPKey{Type: PHPStringKey, Str: key})
-	arr.values = append(arr.values, value)
-}
-
-// Append adds a value to the end of the array with the next available integer key
-func (arr *Array) Append(value interface{}) {
-	nextKey := arr.getNextIntKey()
-	arr.SetInt(nextKey, value)
-}
-
-// getNextIntKey finds the next available integer key
-func (arr *Array) getNextIntKey() int64 {
-	maxKey := int64(-1)
-	for _, key := range arr.keys {
-		if key.Type == PHPIntKey && key.Int > maxKey {
-			maxKey = key.Int
-		}
-	}
-
-	return maxKey + 1
-}
-
-// Len returns the number of elements in the array
-func (arr *Array) Len() uint32 {
-	return uint32(len(arr.keys))
-}
-
-// At returns the key and value at the given index
-func (arr *Array) At(index uint32) (PHPKey, interface{}) {
-	if index >= uint32(len(arr.keys)) {
-		return PHPKey{}, nil
-	}
-	return arr.keys[index], arr.values[index]
-}
-
-// EXPERIMENTAL: GoArray converts a zend_array to a Go Array
-func GoArray(arr unsafe.Pointer) *Array {
-	result := &Array{
-		keys:   make([]PHPKey, 0),
-		values: make([]interface{}, 0),
-	}
+// PackedArray represents a 'packed' PHP array as an indexed list
+type PackedArray = []interface{}
 
+// EXPERIMENTAL: GoAssociativeArray converts a zend_array to a Go AssociativeArray
+func GoAssociativeArray(arr unsafe.Pointer, ordered bool) (AssociativeArray, error) {
 	if arr == nil {
-		return result
+		return AssociativeArray{}, errors.New("GoAssociativeArray received a nil pointer")
 	}
 
 	zval := (*C.zval)(arr)
 	hashTable := (*C.HashTable)(castZval(zval, C.IS_ARRAY))
 
 	if hashTable == nil {
-		return result
+		return AssociativeArray{}, errors.New("GoAssociativeArray received a zval that was not a HashTable")
+	}
+
+	nNumUsed := hashTable.nNumUsed
+	result := AssociativeArray{Map: make(map[string]interface{})}
+	if ordered {
+		result.Order = make([]string, 0, nNumUsed)
 	}
 
-	used := hashTable.nNumUsed
 	if htIsPacked(hashTable) {
-		for i := C.uint32_t(0); i < used; i++ {
+		// if the HashTable is packed, convert all integer keys to strings
+		for i := C.uint32_t(0); i < nNumUsed; i++ {
 			v := C.get_ht_packed_data(hashTable, i)
 			if v != nil && C.zval_get_type(v) != C.IS_UNDEF {
-				value := convertZvalToGo(v)
-				result.SetInt(int64(i), value)
+				strIndex := strconv.Itoa(int(i))
```

Comment:
Couldn't you just use `append()` instead of doing this string to int conversion? Packed arrays are ordered and sequential.